### ISSUE TITLE
Upside down /move subtitle with ISR in fullscreen

### ISSUE THREAD

**Issue by animeavi** [clsid2/mpc-hc#2161]
Cut a sample that reproduces the issue for me, I can reproduce it all the way back to the first "release" in this repo.

Whenever I go fullscreen with that subtitle, it's upside down, switching out of fullscreen makes it fix itself and going back to fullscreen makes it break again. mpv and VLC don't have this issue.

**Edit**: I don't even have to go fullscreen, just resizing the window to be bigger does it.

**Details**: EVR CP, ISR, Windows 11, 1920x1080, GTX 1060

Here's what it looks like when it's broken: 
![upside-down-mpchc](https://github.com/clsid2/mpc-hc/assets/14946679/b0f76e08-b17b-4aa7-9d8f-3f5c70dd3bf1)

MKV with the ASS sub merged in: [upside-down-test.zip](https://github.com/clsid2/mpc-hc/files/11779292/upside-down-test.zip)

ASS sub by itself: [upside-down-sub.zip](https://github.com/clsid2/mpc-hc/files/11779298/upside-down-sub.zip)

**Comment by v0lt** [comment]
I was able to reproduce the issue.
This is a problem with the built-in subtitle renderer. As far as I understand the problem occurs when a certain text size is exceeded. This is due to GDI limitations.

I will advise you to try [XySubFilter with libass](https://github.com/Masaiki/xy-VSFilter/releases).

**Comment by animeavi** [comment]
> I was able to reproduce the issue. This is a problem with the built-in subtitle renderer. As far as I understand the problem occurs when a certain text size is exceeded. This is due to GDI limitations.
> 
> I will advise you to try [XySubFilter with libass](https://github.com/Masaiki/xy-VSFilter/releases).

Thanks for the confirmation.
I do use that, but I always test subs with ISR since most people will probably be using it.

**Comment by adipose** [comment]
![image](https://github.com/clsid2/mpc-hc/assets/3324395/1a340e07-4773-491a-92a6-b19b868aa50f)

This is with the libass capability enabled.

I bet the freetype codepath would also solve this. But that path is pretty limited usefulness--if somehow I could detect it was "too big" I could likely switch over for limited use cases like this.

**Comment by adipose** [comment]
New test builds here for libass build: https://github.com/clsid2/mpc-hc/pull/2163

**Comment by clsid2** [comment]
Decreasing font size and adjusting font scale accordingly could be a solution. Question is if there is a good heuristic to know when size is too big for GDI. Might differ depending on the font?

**Comment by adipose** [comment]
> Decreasing font size and adjusting font scale accordingly could be a solution. Question is if there is a good heuristic to know when size is too big for GDI. Might differ depending on the font?

We could also set a minimum size for GDI scale and then scale the points ourselves.  It would lose some smoothness at the high end.

**Comment by clsid2** [comment]
Yes that is what I meant.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/2169

Magic number is 40000.  At 40000, GetTextExtentPoint32W returns a size of [0,0].  At 40001, it starts to return negative values instead.  Font render is also reversed, of course, but due to negative lengths in GetTextExtentPoint32W, it only flips vertically.

Bit of an ugly hack to add to the code, but it works...

**Comment by v0lt** [comment]
I can't test in MPC-HC, but from my tests in MPC-BE this solution might not work with some fonts.
[upside-down-test-2.zip](https://github.com/clsid2/mpc-hc/files/11885464/upside-down-test-2.zip)

**Comment by clsid2** [comment]
@adipose
I confirm issue with above sample. Text is not upside down, but spacing between fonts increases (too much) as font size increases.

**Comment by adipose** [comment]
> @adipose I confirm issue with above sample. Text is not upside down, but spacing between fonts increases (too much) as font size increases.

Yes, I see it as well.

**Comment by adipose** [comment]
> > @adipose I confirm issue with above sample. Text is not upside down, but spacing between fonts increases (too much) as font size increases.
> 
> Yes, I see it as well.

So I discovered this issue is happening when freetype is enabled.  It seems ok when using GDI only.

**Comment by clsid2** [comment]
If I change
`Style: Default,Arial,400,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2.2,1.1,2,12,12,36,1`
to
`Style: Default,Arial,200,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,200,200,0,0,1,2.2,1.1,2,12,12,36,1`
I get same result. However, second one is still displayed correctly when video is double size.

You previously said this workaround would give wrong positioning. Any details what exactly went wrong?

**Comment by adipose** [comment]
> If I change `Style: Default,Arial,400,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2.2,1.1,2,12,12,36,1` to `Style: Default,Arial,200,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,200,200,0,0,1,2.2,1.1,2,12,12,36,1` I get same result. However, second one is still displayed correctly when video is double size.
> 
> You previously said this workaround would give wrong positioning. Any details what exactly went wrong?

Do you mean this:

```
if (style.fontSize > 39999) {
    int fact = style.fontSize / 39999;
    style.fontSize = 39999;
    style.fontScaleX *= fact;
    style.fontScaleY *= fact;
}
```

This method of scaling the font caused it to get offset incorrectly.  The committed code works fine, in my testing, but it does have issues if you go through the freetype path.  

I am not sure what the issue with freetype path is.  It only happens when doing the post scaling, which we shouldn't really need to do with FT, except for the hack that it tries to match the MM_TEXT scaling measured by `GetTextExtentPoint32W`.

**Comment by Aleksoid1978** [comment]
> > If I change `Style: Default,Arial,400,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2.2,1.1,2,12,12,36,1` to `Style: Default,Arial,200,&H00A44069,&H00A44069,&H00000000,&H00000000,0,0,0,0,200,200,0,0,1,2.2,1.1,2,12,12,36,1` I get same result. However, second one is still displayed correctly when video is double size.
> > You previously said this workaround would give wrong positioning. Any details what exactly went wrong?
> 
> Do you mean this:
> 
> ```
> if (style.fontSize > 39999) {
>     int fact = style.fontSize / 39999;
>     style.fontSize = 39999;
>     style.fontScaleX *= fact;
>     style.fontScaleY *= fact;
> }
> ```
> 
> This method of scaling the font caused it to get offset incorrectly. The committed code works fine, in my testing, but it does have issues if you go through the freetype path.
> 
> I am not sure what the issue with freetype path is. It only happens when doing the post scaling, which we shouldn't really need to do with FT, except for the hack that it tries to match the MM_TEXT scaling measured by `GetTextExtentPoint32W`.

```
int fact = style.fontSize / 39999;
style.fontScaleX *= fact;
style.fontScaleY *= fact;
```
I think this is not correct due to rounding when casting to int, it should probably be like this:
```
auto fact = static_cast<double>(style.fontSize) / 39999.;
style.fontScaleX *= fact;
style.fontScaleY *= fact;
```

**Comment by adipose** [comment]
Yikes, that was the whole problem...

https://github.com/clsid2/mpc-hc/pull/2171

This does indeed work fine and is much simpler.

**Comment by adipose** [comment]
This also seems to solve the freetype + largefont issue that caused funny spacing.  I think the problem there was the freetype codepath was using the modified `GetTextExtent` which was confusing the freetype render for targeted font size.

**Comment by clsid2** [comment]
What is the reason for MPCBE using smaller value?
[MAXGDIFONTSIZE 15087](https://github.com/Aleksoid1978/MPC-BE/commit/fb043c4000d433d62a473638300798906f3480b6)

**Comment by Aleksoid1978** [comment]
> What is the reason for MPCBE using smaller value? [MAXGDIFONTSIZE 15087](https://github.com/Aleksoid1978/MPC-BE/commit/fb043c4000d433d62a473638300798906f3480b6)

Because with your number on some fonts the text is upside down anyway.
Check all subtitles from [upside-down-test-2.zip](https://github.com/clsid2/mpc-hc/files/11885464/upside-down-test-2.zip)

**Comment by adipose** [comment]
I also wondered why MPC-HC uses MM_TEXT which is supposed to use negative font sizes, but then uses positive font sizes instead.  But in the negative realm, the limit is more like 32767 instead of 40000.

40000 happens to correspond to a 300 point font @96dpi.  Not sure how 15087 is calculated.

**Comment by adipose** [comment]
> > What is the reason for MPCBE using smaller value? [MAXGDIFONTSIZE 15087](https://github.com/Aleksoid1978/MPC-BE/commit/fb043c4000d433d62a473638300798906f3480b6)
> 
> Because with your number on some fonts the text is upside down anyway. Check all subtitles from [upside-down-test-2.zip](https://github.com/clsid2/mpc-hc/files/11885464/upside-down-test-2.zip)

In MPC-HC 39999 is working on this test, at least at 192dpi full screen.

**Comment by clsid2** [comment]
Only Calibri and ComisSans are good here on 200% video size. With 15000 all are good.

**Comment by adipose** [comment]
> Only Calibri and ComisSans are good here on 200% video size. With 15000 all are good.

I see it now, my freetype path was enabled.

**Comment by adipose** [comment]
So I have tested all those samples and I cannot find any that have issues beyond 32767.  I have observed that there are two fail points:

1. `GetTextExtentPoint32W` fails at 40000 fontsize (or -32768)
2. `DrawTextW` fails at different points between 32768 and 50000 (Calibri: 40000, Arial: ~36750, Comic Sans: ~46000, Consolas: ~38500, Times NR ~36000).

Since either one will mess up the rendering we should choose the lowest.  So far, I didn't find any font below 36000 that had trouble.  I wonder if this has to do with escapement...
